xc/device.ml

```diff
@@ -607,12 +607,35 @@ let qemu_media_change ~xs device _type params =
 
 	if is_upstream_qemu device.frontend.domid
 	then begin
+		let open Qmp in
 		let cd = "ide1-cd1" in
-		let qmp_cmd =
-			if params = ""
-			then (Qmp.Command(None, Qmp.Eject (cd, Some true)))
-			else (Qmp.Command(None, Qmp.Change (cd, params, None))) in
-		qmp_write device.frontend.domid qmp_cmd
+		match params with
+		| "" ->
+			let qmp_cmd = Command(None, Eject (cd, Some true)) in
+			qmp_write device.frontend.domid qmp_cmd
+		| _ ->
+			try
+				let c = Qmp_protocol.connect (Printf.sprintf "/var/run/xen/qmp-libxl-%d" device.frontend.domid) in
+				let fd_of_c = Qmp_protocol.to_fd c in
+				let fd_of_cd = Unix.openfile params [ Unix.O_RDONLY ] 0o640 in
+				finally
+					(fun () -> ignore(Fd_send_recv.send_fd fd_of_c " " 0 1 [] fd_of_cd))
+					(fun () -> Unix.close fd_of_cd);
+				(* It' ok here to use a closed fd_of_cd *)
+				let qmp_cmd = Command (None, Add_fd (Fd_send_recv.int_of_fd fd_of_cd)) in
```

Comment:
I think this should be `Add_fd (None)` or similar to return a fresh, unused fdset_id associated with the fd_of_cd that has just been sent via SCM_RIGHTS. This is what the add-fd documentation says:
https://github.com/dhruvvyas90/qemu/blob/master/qmp-commands.hx#L1743
> add-fd
> -------
> Add a file descriptor, that was passed via SCM rights, to an fd set.
> Arguments: (optional)
> ...
> (2) If "fdset-id" is not specified, a new fd set will be created.

I understand from the above that qemu will keep a list of fds passed via SCM, and when add-fd is called without parameters, a new fdset_id will be created, containing the fds passed via SCM since the last add-fd was called. The resulting fdset_id and fd can be verified by sending query-fdsets from a terminal: https://github.com/dhruvvyas90/qemu/blob/master/qmp-commands.hx#L1814
Passing Add_fd (Fd_send_recv.int_of_fd fd_of_cd) seems wrong to me because it's an unpredictable int that could return the same id or different id on different calls for different clients of this qemu processes -- it could for instance clash the fdset for different calls of this function or with different qmp clients of this qemu process, such as libxl. 

@idruzhinin do you think the above that makes sense?